### PR TITLE
HPCC-31661 Add options to generate cpu resources as limits

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
From Copilot (experimental)

This pull request primarily focuses on the modification of the Helm charts in the `hpcc` package. The changes include the addition of new global parameters to manage resource limits and requests, and the refactoring of the resource allocation code across various templates. The new parameters, `resourceCpusWithLimits` and `resourceWholeCpusWithLimits`, provide more flexibility in setting CPU resource limits. 

Key changes include:

Addition of new global parameters:

* [`helm/hpcc/values.schema.json`](diffhunk://#diff-362d00cdb2c372ff1cb532bbec5653be6039d90f5d1ee0cbc133d8b94e3a4489R276-R283): Added new global parameters `resourceCpusWithLimits` and `resourceWholeCpusWithLimits` to manage CPU resource limits.

Refactoring of resource allocation code:

* [`helm/hpcc/templates/_helpers.tpl`](diffhunk://#diff-3d8bff06a9eeeb48fccd515e43fac425b9cabb2fc735e237150ff2827d70ab6fR1319-R1340): Modified the `hpcc.addResources` helper to use the new global parameters and refactored the resource allocation code. [[1]](diffhunk://#diff-3d8bff06a9eeeb48fccd515e43fac425b9cabb2fc735e237150ff2827d70ab6fR1319-R1340) [[2]](diffhunk://#diff-3d8bff06a9eeeb48fccd515e43fac425b9cabb2fc735e237150ff2827d70ab6fR1352-R1363)
* Various `helm/hpcc/templates/*.yaml` files: Refactored the resource allocation code to use the `hpcc.addResources` helper with the new `root` parameter. [[1]](diffhunk://#diff-0bfca39aff0006c66b706c4b7e5d00a13ce729b7cefb3532ac8c067ae5f31851L77-R77) [[2]](diffhunk://#diff-da89055d1f1aaedc1cf2eff813658a2b5df05656080db0302836adce00aac974L119-R119) [[3]](diffhunk://#diff-b75c439b73064c7018a310825e39130b4c3793833207be6333bbd67c26d3614fL87-R87) [[4]](diffhunk://#diff-738e0e9649cb63234beadc210636d7407d3c344a9e17e66bf555533703993798L85-R85) [[5]](diffhunk://#diff-738e0e9649cb63234beadc210636d7407d3c344a9e17e66bf555533703993798L170-R170) [[6]](diffhunk://#diff-d2f1c0870590a96cbe1a5d626b5c82a286e9f7546a07b5ba40f5ad0efc4492adL80-R80) [[7]](diffhunk://#diff-d2f1c0870590a96cbe1a5d626b5c82a286e9f7546a07b5ba40f5ad0efc4492adL178-R179) [[8]](diffhunk://#diff-f9e0a09fa408c0c3d1bbdc7bdbf67ec917d56cb2de6c8dafcec5cc3122cdbdf1L96-R97) [[9]](diffhunk://#diff-78d101699abf01e75c93f8c40a864dcfe0a02bb1c15803a4784d47dee5d9988bL158-R158) [[10]](diffhunk://#diff-d48e1a5ec275ace54e4be1434b3a414f73a729c4c92629c14ae0a73c2d693140L117-R117) [[11]](diffhunk://#diff-85a06e251822e499cf1f7f34ef4ef08b66842f319f49739d105fdf466c6a1c88L112-R112) [[12]](diffhunk://#diff-85a06e251822e499cf1f7f34ef4ef08b66842f319f49739d105fdf466c6a1c88L182-R179) [[13]](diffhunk://#diff-85a06e251822e499cf1f7f34ef4ef08b66842f319f49739d105fdf466c6a1c88L254-R248) [[14]](diffhunk://#diff-85a06e251822e499cf1f7f34ef4ef08b66842f319f49739d105fdf466c6a1c88L393-R387) [[15]](diffhunk://#diff-85a06e251822e499cf1f7f34ef4ef08b66842f319f49739d105fdf466c6a1c88L459-R447)

These changes are aimed at providing more flexibility in managing CPU resources in a Kubernetes environment, especially when the `cpuManagerPolicy=static` is used, which allows binding using affinities.

File: helm/hpcc/templates/_helpers.tpl [hunk 1]
```diff
@@ -1316,18 +1316,38 @@ Pass in a dictionary with me defined
 */}}
 {{- define "hpcc.addResources" }}
 {{- $resources := .me | default .defaults }}
-{{- if $resources }}
- {{- $limits := omit $resources "cpu" }}
- {{- $requests := pick $resources "cpu" }}
+{{- $omitResources := hasKey .root.Values.global "omitResources" | ternary .root.Values.global.omitResources .root.Values.global.privileged }}
+{{- $resourceCpusWithLimits := hasKey .root.Values.global "resourceCpusWithLimits" | ternary .root.Values.global.resourceCpusWithLimits  false -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1320]: double space before 'false'

File: helm/hpcc/templates/_helpers.tpl [hunk 2]
```diff
@@ -1316,18 +1316,38 @@ Pass in a dictionary with me defined
...
+{{- $resourceWholeCpusWithLimits := hasKey .root.Values.global "resourceWholeCpusWithLimits" | ternary .root.Values.global.resourceWholeCpusWithLimits  false -}}
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1321]: trivial: double space before 'false'

**Review by jakesmith — APPROVED** [review]
@ghalliday - 2 trivial spacing issue. Looks good.

**Comment by ghalliday** [comment]
@jakesmith please re-review for sanity.  I have rebased onto latest to include your privileged change and the increase in stub resources.  I have sanity checked all the parameters are correct, but looking for a second opinion to check we don't break something.

File: helm/hpcc/values.schema.json [hunk 3]
```diff
@@ -273,6 +273,14 @@
           "description": "if set, no resource definitions are generated from the helm charts",
           "type": "boolean"
         },
+        "resourceCpusWithLimits": {
+          "description": "if set, cpu resources are provided as limits rather than requests.  Not recommended because this can lead to painful throttling on roxie",
```
Comment (jakesmith) [inline helm/hpcc/values.schema.json:277]: not just roxie - we saw on Thor originally I think

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good afaics. 1 trivial comment.

**Pull request merged**